winpty/src/windows.rs

```diff
@@ -260,36 +282,29 @@ impl SpawnConfig {
         cmdline: Option<&str>,
         cwd: Option<&str>,
         end: Option<&str>,
-    ) -> Result<Self, Err> {
+    ) -> Result<Self, Error> {
         let mut err = null_mut() as *mut winpty_error_t;
-        let (appname, cmdline, cwd, end) = (
-            appname.map_or(null(), |s| WideCString::from_str(s).unwrap().into_raw()),
-            cmdline.map_or(null(), |s| WideCString::from_str(s).unwrap().into_raw()),
-            cwd.map_or(null(), |s| WideCString::from_str(s).unwrap().into_raw()),
-            end.map_or(null(), |s| WideCString::from_str(s).unwrap().into_raw()),
-        );
 
+        let to_wstring = |s| WideCString::from_str(s).unwrap();
+        let appname = appname.map(to_wstring);
+        let cmdline = cmdline.map(to_wstring);
+        let cwd = cwd.map(to_wstring);
+        let end = end.map(to_wstring);
+
+        let wstring_ptr = |opt: &Option<WideCString>| opt.as_ref().map_or(null(), |ws| ws.as_ptr());
         let spawn_config = unsafe {
-            winpty_spawn_config_new(spawnflags.bits(), appname, cmdline, cwd, end, &mut err)
+            winpty_spawn_config_new(
+                spawnflags.bits(),
+                wstring_ptr(&appname),
+                wstring_ptr(&cmdline),
+                wstring_ptr(&cwd),
+                wstring_ptr(&end),
+                &mut err,
+            )
         };
 
-        // Required to free the strings
-        unsafe {
-            if !appname.is_null() {
-                WideCString::from_raw(appname as *mut u16);
-            }
-            if !cmdline.is_null() {
-                WideCString::from_raw(cmdline as *mut u16);
-            }
-            if !cwd.is_null() {
-                WideCString::from_raw(cwd as *mut u16);
-            }
-            if !end.is_null() {
-                WideCString::from_raw(end as *mut u16);
-            }
-        }
-
-        check_err(err).map_or(Ok(Self(spawn_config)), Result::Err)
+        check_err(err)?;
+        Ok(Self(spawn_config))
```

Comment:
Could we add an extra newline between the three instances where `check_err` now does an early return and we explicitly return `Ok(...)` ourselves?

Since the two aren't related anymore, but two completely separate operations, I'd like to separate that semantically.

Otherwise it would make more sense to use `check_err(err).map(|_| ...)`, but I think the idea of not mapping random data into an error is good.